### PR TITLE
Add Catchment-scale Simulation

### PR DESCRIPTION
**Pull request by leoglonz**
## Issue Addressed

Add new script `scripts/sim.py` and capability in mc to forward model and output routed streamflow at the scale of individual catchments.

## Description

Per above, adds new script for simulation only (similar to `scripts/test.py` excluding metric evaluation), a new boolean option to config `experiment.catchment_out` and `src/routing/ddr/mmc.py` which, if true, will direct mmc to output catchment-level routed streamflow instead of summing to gauges. 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code cleanup/refactor
- [ ] Documentation update

Other (please specify):

## Checklist

- [x] Branch is up to date with master
- [ ] Updated tests or added new tests
- [ ] Tests & pre-commit hooks pass
- [ ] Updated documentation (if applicable)
- [x] Code follows established style and conventions

### CODE REVIEW

File: examples/eval/evaluate.ipynb
Comment (taddyb) [inline examples/eval/evaluate.ipynb:1]: need to get rid of the hardcoded paths

**Review by taddyb — COMMENTED** [review]
I think  just changing the output here in `mmc.py` is necessary
```
                # Store output at gauge locations
                for i, gage_idx in enumerate(self.gage_indices):
                    output[i, timestep] = torch.sum(q_t1[gage_idx])
```

File: examples/eval/evaluate.ipynb
Comment (leoglonz) [reply examples/eval/evaluate.ipynb:1]: Actually, yeah, did not intend to include this. Will remove

**Comment by taddyb** [comment]
As an update to this PR, I introduced the `target_catchments` data source inside of the config file. It's a new feature that will route for everything upstream of that `target_catchment` and return the output for only those sections of the hydrofabric. 

example:
your config would be
```yaml
data_sources:
  attributes: /projects/mhpi/data/icechunk/hydrofabric_v2.2_attributes
  hydrofabric_gpkg: /projects/mhpi/tbindas/ddr/data/conus_nextgen.gpkg  # patched
  conus_adjacency: /projects/mhpi/tbindas/ddr/data/conus_adjacency.zarr
  # gages_adjacency: /projects/mhpi/tbindas/ddr/data/_gages_adjacency.zarr
  statistics: /projects/mhpi/tbindas/ddr/data
  streamflow: /projects/mhpi/data/icechunk/hydrofabric_v2.2_dhbv_retrospective
  observations: /projects/mhpi/data/icechunk/usgs_daily_observations
  # gages: /projects/mhpi/tbindas/datasets/mhpi/dHBV2.0UH/eval_test_3.csv
  # gages: /projects/mhpi/tbindas/datasets/mhpi/dHBV2.0UH/training_gauges.csv
  target_catchments:
    - wb-86522
    - wb-6577
```
and you would not have a gages or gage_adjacency reference. This code would then route for all flows upstream of those flowpaths (see photo below) for all of the timesteps specified in your config. This is ON DEMAND and just requires you have the CONUS gpkg file in a coo-zarr format

image

This is all within the `scripts/router.py` file

#### NOTE
I also managed to cut runtime in half by introducing a `scatter_add` that maps discharge to the output array and removed the for loop that was there before. The change is replacing 
```py
            # Store output at gauge locations
            for i, gage_idx in enumerate(self.gage_indices):
                output[i, timestep] = torch.sum(q_t1[gage_idx])
```

with 
```py
    def setup_inputs()
        .
        .
        .
        # Precompute scatter_add indices for ragged output_indices (gages mode)
        if self.output_indices is not None and len(self.output_indices) != len(self._discharge_t):
            self._flat_indices = torch.cat(
                [torch.as_tensor(idx, device=self.device, dtype=torch.long) for idx in self.output_indices]
            )
            self._group_ids = torch.cat(
                [
                    torch.full((len(idx),), i, device=self.device, dtype=torch.long)
                    for i, idx in enumerate(self.output_indices)
                ]
            )
            self._num_outputs = len(self.output_indices)
            self._scatter_input = torch.zeros(self._num_outputs, device=self.device, dtype=torch.float32)
        else:
            self._flat_indices = None
            self._group_ids = None
            self._num_outputs = None
            self._scatter_input = None

    def forward()
        .
        .
        .
            if output_all:
                output[:, timestep] = q_t1
            else:
                gathered = q_t1[self._flat_indices]
                output[:, timestep] = torch.scatter_add(
                    input=self._scatter_input,
                    dim=0,
                    index=self._group_ids,
                    src=gathered,
                )
```
this scatter add replaces the torch.sum() and has been checked through training by verifying NSE, loss, and time-series output across all gages

it improves training speed substantially! By ~50%
